File: internal/detection/face_detector.go
package detection

import (
	"image"
	"math"

	"gocv.io/x/gocv"
)

// smoothAlpha controls EMA smoothing: closer to 1.0 = more responsive, closer to 0.0 = smoother.
const smoothAlpha = 0.35

// minMotionSize is the minimum width/height (px) for a motion region to be considered.
const minMotionSize = 10

// minIoU is the minimum Intersection-over-Union to match a new detection to a tracked face.
const minIoU = 0.15

// ── Rolling-average filter constants ────────────────────────────────────────

// areaHistoryLen is the number of past frames used to compute the rolling
// average face area. Larger values give a more stable baseline but react
// more slowly to genuine size changes (e.g. person walking closer).
const areaHistoryLen = 15

// minHistoryForFilter is the minimum number of frames that must be in the
// history buffer before area- or jump-based filtering kicks in.
// During early frames the filter is intentionally skipped so the tracker
// can establish a baseline quickly.
const minHistoryForFilter = 5

// maxAreaRatio is the maximum allowed ratio between a candidate's area and
// the rolling average. A detection is rejected when:
//
//	candidate_area > avg * maxAreaRatio  (sudden enlargement)
//	candidate_area < avg / maxAreaRatio  (sudden shrinkage)
const maxAreaRatio = 2.5

// maxCenterJumpFrac is the maximum allowed displacement of the face centre
// expressed as a fraction of the frame diagonal. Detections whose centre
// jumps farther than this in a single frame are considered spurious.
const maxCenterJumpFrac = 0.20

// absentTimeoutFrames is the number of consecutive frames with no valid face
// detection before the detector declares that nobody is in front of the camera.
// At 30 fps, 60 frames ≈ 2 seconds. Increase for a longer grace period.
const absentTimeoutFrames = 60

// ─────────────────────────────────────────────
// TrackedFace – a single tracked face instance
// ─────────────────────────────────────────────

// fRect is a float64 version of image.Rectangle for smooth interpolation.
type fRect struct{ x, y, x2, y2 float64 }

func toFRect(r image.Rectangle) fRect {
	return fRect{
		x:  float64(r.Min.X),
		y:  float64(r.Min.Y),
		x2: float64(r.Max.X),
		y2: float64(r.Max.Y),
	}
}

func (f fRect) toRect() image.Rectangle {
	return image.Rect(int(math.Round(f.x)), int(math.Round(f.y)),
		int(math.Round(f.x2)), int(math.Round(f.y2)))
}

// lerp applies EMA to each coordinate.
func (f fRect) lerp(target image.Rectangle, alpha float64) fRect {
	t := toFRect(target)
	return fRect{
		x:  f.x + alpha*(t.x-f.x),
		y:  f.y + alpha*(t.y-f.y),
		x2: f.x2 + alpha*(t.x2-f.x2),
		y2: f.y2 + alpha*(t.y2-f.y2),
	}
}

type trackedFace struct {
	smooth  fRect // smoothed (display) rect
	matched bool  // was it matched this frame?
}

// ─────────────────────────────────────────────
// FaceDetector
// ─────────────────────────────────────────────

// FaceDetector detects and tracks a single primary face across frames.
// It uses frame-differencing to narrow the search region, a Haar-cascade
// classifier for detection, IoU matching for continuity, and EMA smoothing
// to avoid jitter.
type FaceDetector struct {
	classifier gocv.CascadeClassifier

	// motion-diff mats
	prevGray gocv.Mat
	gray     gocv.Mat
	diffMat  gocv.Mat
	thresh   gocv.Mat

	// tracked state
	track *trackedFace

	// frame dimensions (updated each Detect call)
	imgW, imgH int

	// ── Rolling-average area filter ──────────────────────────────────────
	// areaHistory is a circular buffer storing rectArea of each accepted
	// face detection over the last areaHistoryLen frames.
	areaHistory [areaHistoryLen]int
	areaIdx     int // next write position in the circular buffer
	areaCount   int // how many valid entries are currently in the buffer

	// ── Absence detection ────────────────────────────────────────────────
	// missFrames counts consecutive frames in which no valid face was found.
	// Once it reaches absentTimeoutFrames the tracker is reset and
	// facePresent is set to false.
	missFrames  int
	facePresent bool // true while a face is considered to be in the frame
}

// NewFaceDetector creates a FaceDetector and loads the Haar cascade from cascadePath.
// Returns an error string (non-empty) if the file cannot be loaded.
func NewFaceDetector(cascadePath string) (*FaceDetector, bool) {
	fd := &FaceDetector{
		classifier: gocv.NewCascadeClassifier(),
		prevGray:   gocv.NewMat(),
		gray:       gocv.NewMat(),
		diffMat:    gocv.NewMat(),
		thresh:     gocv.NewMat(),
	}
	if !fd.classifier.Load(cascadePath) {
		fd.Close()
		return nil, false
	}
	return fd, true
}

// Close releases all OpenCV resources held by FaceDetector.
func (fd *FaceDetector) Close() {
	fd.classifier.Close()
	fd.prevGray.Close()
	fd.gray.Close()
	fd.diffMat.Close()
	fd.thresh.Close()
}

// Detect analyses img and returns the smoothed bounding rectangle of the
// primary (largest / closest) face, plus a boolean indicating whether a
// face is currently being tracked.
//
// After absentTimeoutFrames consecutive frames with no valid detection the
// tracker resets and IsPresent() returns false.
func (fd *FaceDetector) Detect(img gocv.Mat) (image.Rectangle, bool) {
	// store frame size for NormalizedPosition
	fd.imgW = img.Cols()
	fd.imgH = img.Rows()

	// ── 1. Compute motion mask ──────────────────────────────────────────
	gocv.CvtColor(img, &fd.gray, gocv.ColorBGRToGray)

	if fd.prevGray.Empty() {
		fd.gray.CopyTo(&fd.prevGray)
		return image.Rectangle{}, false
	}

	gocv.AbsDiff(fd.gray, fd.prevGray, &fd.diffMat)
	gocv.Threshold(fd.diffMat, &fd.thresh, 25, 255, gocv.ThresholdBinary)

	kernel := gocv.GetStructuringElement(gocv.MorphRect, image.Pt(15, 15))
	gocv.Dilate(fd.thresh, &fd.thresh, kernel)
	kernel.Close()

	fd.gray.CopyTo(&fd.prevGray)

	motionRect := motionBoundingRect(fd.thresh)
	hasMotion := motionRect.Dx() > minMotionSize && motionRect.Dy() > minMotionSize

	// ── 2. Run classifier in motion ROI ────────────────────────────────
	var rawDetections []image.Rectangle

	if hasMotion {
		// Clamp ROI to image bounds
		imgBounds := image.Rect(0, 0, img.Cols(), img.Rows())
		roi2 := motionRect.Intersect(imgBounds)

		if !roi2.Empty() {
			roiMat := img.Region(roi2)
			detected := fd.classifier.DetectMultiScale(roiMat)
			roiMat.Close()

			// Translate back to full-image coordinates
			for _, f := range detected {
				rawDetections = append(rawDetections, image.Rect(
					roi2.Min.X+f.Min.X, roi2.Min.Y+f.Min.Y,
					roi2.Min.X+f.Max.X, roi2.Min.Y+f.Max.Y,
				))
			}
		}
	} else if fd.track != nil {
		// No motion – keep last smoothed position but still tick the miss
		// counter so a perfectly still person who walked away is eventually
		// timed out (motion simply drops to zero once they're gone).
		// We do NOT increment missFrames here because a static scene is
		// indistinguishable from a still face; only a full absence of a
		// detection inside a motion region counts as a miss.
		return fd.track.smooth.toRect(), true
	}

	// ── 3. Pick largest detection (closest to camera) ──────────────────
	best := largestRect(rawDetections)

	// ── 4. No detection or failed plausibility ──────────────────────────
	// Handle the "no face found" path first so absence tracking is in one place.
	if best.Empty() || !fd.isPlausibleDetection(best) {
		// Increment miss counter.
		fd.missFrames++

		if fd.missFrames >= absentTimeoutFrames {
			// Timeout reached – declare nobody present and wipe all state
			// so the detector starts fresh when a face reappears.
			fd.track = nil
			fd.facePresent = false
			fd.areaCount = 0
			fd.areaIdx = 0
			return image.Rectangle{}, false
		}

		// Within grace period – keep the last known position if available.
		if fd.track != nil {
			return fd.track.smooth.toRect(), true
		}
		return image.Rectangle{}, false
	}

	// ── 5. Valid detection – update tracker ─────────────────────────────
	// Reset miss counter and mark face as present.
	fd.missFrames = 0
	fd.facePresent = true

	if fd.track == nil {
		// First plausible detection – initialise tracker.
		fd.track = &trackedFace{smooth: toFRect(best)}
	} else {
		// IoU matching + EMA smoothing.
		iou := iouRect(fd.track.smooth.toRect(), best)
		if iou >= minIoU {
			fd.track.smooth = fd.track.smooth.lerp(best, smoothAlpha)
		} else {
			fd.track.smooth = toFRect(best)
		}
	}

	// Record accepted area into the rolling history.
	fd.pushArea(rectArea(best))

	return fd.track.smooth.toRect(), true
}

// MotionBoundingRect is exported for callers that want to draw the motion region.
func (fd *FaceDetector) MotionBoundingRect(img gocv.Mat) image.Rectangle {
	return motionBoundingRect(fd.thresh)
}

// NormalizedPosition returns the centre of the tracked face as (x, y) values
// in the range [0, 1], where:
//
//	(0, 0) = top-left corner of the frame
//	(1, 1) = bottom-right corner of the frame
//
// ok is false when no face is currently tracked or the frame size is unknown.
func (fd *FaceDetector) NormalizedPosition() (x, y float64, ok bool) {
	if fd.track == nil || fd.imgW == 0 || fd.imgH == 0 {
		return 0, 0, false
	}
	r := fd.track.smooth
	cx := (r.x + r.x2) / 2.0
	cy := (r.y + r.y2) / 2.0
	x = math.Max(0, math.Min(1, cx/float64(fd.imgW)))
	y = math.Max(0, math.Min(1, cy/float64(fd.imgH)))
	return x, y, true
}

// IsPresent returns true if a face has been detected and the absence timeout
// has not yet elapsed. This is the primary signal for "is someone in front of
// the camera right now?"
func (fd *FaceDetector) IsPresent() bool {
	return fd.facePresent
}

// MissFrames returns the number of consecutive frames in which no valid face
// detection was found. Resets to 0 on every accepted detection.
// Callers can use this to display a countdown or drive application logic
// before the full absentTimeoutFrames threshold is reached.
func (fd *FaceDetector) MissFrames() int {
	return fd.missFrames
}

// ─────────────────────────────────────────────
// Rolling-average area filter
// ─────────────────────────────────────────────

// pushArea records area into the circular history buffer.
func (fd *FaceDetector) pushArea(area int) {
	fd.areaHistory[fd.areaIdx] = area
	fd.areaIdx = (fd.areaIdx + 1) % areaHistoryLen
	if fd.areaCount < areaHistoryLen {
		fd.areaCount++
	}
}

// RollingAvgArea returns the rolling average face area (px²) over the last
// areaHistoryLen accepted detections, and the number of samples used.
// If no samples are available yet, returns (0, 0).
func (fd *FaceDetector) RollingAvgArea() (avg float64, n int) {
	n = fd.areaCount
	if n == 0 {
		return 0, 0
	}
	sum := 0
	for i := 0; i < n; i++ {
		sum += fd.areaHistory[i]
	}
	return float64(sum) / float64(n), n
}

// isPlausibleDetection returns true when candidate passes the area-ratio and
// centre-jump plausibility checks against the rolling history.
//
// It returns true unconditionally until minHistoryForFilter accepted frames
// have been accumulated, so the tracker can bootstrap without being blocked.
func (fd *FaceDetector) isPlausibleDetection(candidate image.Rectangle) bool {
	if fd.areaCount < minHistoryForFilter {
		// Not enough history yet – let anything through.
		return true
	}

	avgArea, _ := fd.RollingAvgArea()
	candArea := float64(rectArea(candidate))

	// ── Area-ratio check ────────────────────────────────────────────────
	// Reject if the candidate is dramatically larger or smaller than the
	// rolling average (both directions use the same ratio threshold).
	if candArea > avgArea*maxAreaRatio {
		return false // suddenly much larger → likely false positive
	}
	if candArea < avgArea/maxAreaRatio {
		return false // suddenly much smaller → likely false positive
	}

	// ── Centre-jump check ───────────────────────────────────────────────
	// Reject if the face centre has moved further than maxCenterJumpFrac
	// of the frame diagonal in a single frame.
	if fd.track != nil && fd.imgW > 0 && fd.imgH > 0 {
		diag := math.Sqrt(float64(fd.imgW*fd.imgW + fd.imgH*fd.imgH))
		maxJump := maxCenterJumpFrac * diag

		// Current smoothed centre
		sr := fd.track.smooth
		scx := (sr.x + sr.x2) / 2.0
		scy := (sr.y + sr.y2) / 2.0

		// Candidate centre
		ccx := float64(candidate.Min.X+candidate.Max.X) / 2.0
		ccy := float64(candidate.Min.Y+candidate.Max.Y) / 2.0

		dx := ccx - scx
		dy := ccy - scy
		jump := math.Sqrt(dx*dx + dy*dy)

		if jump > maxJump {
			return false // centre teleported → spurious detection
		}
	}

	return true
}

// ─────────────────────────────────────────────
// Helpers (package-level, shared)
// ─────────────────────────────────────────────

// motionBoundingRect returns the union of all contour bounding rects in mask.
func motionBoundingRect(mask gocv.Mat) image.Rectangle {
	contours := gocv.FindContours(mask, gocv.RetrievalExternal, gocv.ChainApproxSimple)
	defer contours.Close()

	if contours.Size() == 0 {
		return image.Rectangle{}
	}

	union := image.Rectangle{}
	for i := 0; i < contours.Size(); i++ {
		br := gocv.BoundingRect(contours.At(i))
		if union.Empty() {
			union = br
		} else {
			union = union.Union(br)
		}
	}
	return union
}

// largestRect returns the rectangle with the largest area from a slice.
func largestRect(rects []image.Rectangle) image.Rectangle {
	var best image.Rectangle
	for _, r := range rects {
		if rectArea(r) > rectArea(best) {
			best = r
		}
	}
	return best
}

// rectArea returns the pixel area of a rectangle.
func rectArea(r image.Rectangle) int {
	return r.Dx() * r.Dy()
}

// iouRect computes Intersection-over-Union for two rectangles.
func iouRect(a, b image.Rectangle) float64 {
	inter := a.Intersect(b)
	if inter.Empty() {
		return 0
	}
	interArea := float64(rectArea(inter))
	unionArea := float64(rectArea(a)+rectArea(b)) - interArea
	if unionArea == 0 {
		return 0
	}
	return interArea / unionArea
}

File: backend/app/lib/face_detector.py
from __future__ import annotations

from dataclasses import dataclass
import math
from pathlib import Path

import cv2
import numpy as np

smooth_alpha = 0.35
min_motion_size = 10
min_iou = 0.15

area_history_len = 15
min_history_for_filter = 5
max_area_ratio = 2.5
max_center_jump_frac = 0.20
absent_timeout_frames = 60


@dataclass(slots=True)
class FloatRect:
    x: float
    y: float
    x2: float
    y2: float

    @classmethod
    def from_rect(cls, rect: tuple[int, int, int, int]) -> "FloatRect":
        x, y, w, h = rect
        return cls(float(x), float(y), float(x + w), float(y + h))

    def to_rect(self) -> tuple[int, int, int, int]:
        x = int(round(self.x))
        y = int(round(self.y))
        x2 = int(round(self.x2))
        y2 = int(round(self.y2))
        return x, y, max(0, x2 - x), max(0, y2 - y)

    def lerp(self, target: tuple[int, int, int, int], alpha: float) -> "FloatRect":
        other = FloatRect.from_rect(target)
        return FloatRect(
            x=self.x + alpha * (other.x - self.x),
            y=self.y + alpha * (other.y - self.y),
            x2=self.x2 + alpha * (other.x2 - self.x2),
            y2=self.y2 + alpha * (other.y2 - self.y2),
        )


class FaceDetector:
    def __init__(self, cascade_path: Path | str) -> None:
        self._classifier = cv2.CascadeClassifier(str(cascade_path))
        if self._classifier.empty():
            raise RuntimeError(f"failed to load haar cascade: {cascade_path}")

        self._prev_gray: np.ndarray | None = None
        self._tracked: FloatRect | None = None
        self._img_w = 0
        self._img_h = 0
        self._area_history: list[int] = []
        self._miss_frames = 0
        self._face_present = False
        self._last_motion_rect: tuple[int, int, int, int] | None = None
        self._full_frame_scan_interval = 2
        self._frame_counter = 0

    def close(self) -> None:
        self._prev_gray = None
        self._tracked = None
        self._area_history.clear()
        self._last_motion_rect = None
        self._miss_frames = 0
        self._face_present = False
        self._frame_counter = 0

    def detect(self, frame: np.ndarray) -> tuple[tuple[int, int, int, int] | None, bool]:
        self._img_h, self._img_w = frame.shape[:2]
        self._frame_counter += 1
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

        if self._prev_gray is None or self._prev_gray.shape != gray.shape:
            self._prev_gray = gray.copy()
            self._last_motion_rect = None
            self._tracked = None
            self._miss_frames = 0
            self._face_present = False
            return None, False

        diff = cv2.absdiff(gray, self._prev_gray)
        _, thresh = cv2.threshold(diff, 25, 255, cv2.THRESH_BINARY)
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (15, 15))
        thresh = cv2.dilate(thresh, kernel, iterations=1)
        self._prev_gray = gray.copy()

        motion_rect = self._motion_bounding_rect(thresh)
        self._last_motion_rect = motion_rect
        has_motion = motion_rect is not None and motion_rect[2] > min_motion_size and motion_rect[3] > min_motion_size

        raw_detections: list[tuple[int, int, int, int]] = []
        if has_motion and motion_rect is not None:
            raw_detections = self._detect_in_region(gray, motion_rect)
        elif self._tracked is not None:
            return self._tracked.to_rect(), True
        elif self._should_scan_full_frame():
            raw_detections = self._detect_in_region(gray, (0, 0, self._img_w, self._img_h))

        if not raw_detections and self._tracked is None and self._should_scan_full_frame():
            raw_detections = self._detect_in_region(gray, (0, 0, self._img_w, self._img_h))

        best = self._largest_rect(raw_detections)
        if best is None or not self._is_plausible_detection(best):
            self._miss_frames += 1
            if self._miss_frames >= absent_timeout_frames:
                self._tracked = None
                self._face_present = False
                self._area_history.clear()
                return None, False
            if self._tracked is not None:
                return self._tracked.to_rect(), True
            return None, False

        self._miss_frames = 0
        self._face_present = True

        if self._tracked is None:
            self._tracked = FloatRect.from_rect(best)
        else:
            iou = self._iou_rect(self._tracked.to_rect(), best)
            if iou >= min_iou:
                self._tracked = self._tracked.lerp(best, smooth_alpha)
            else:
                self._tracked = FloatRect.from_rect(best)

        self._push_area(best[2] * best[3])
        return self._tracked.to_rect(), True

    def normalized_position(self) -> tuple[float, float, bool]:
        if self._tracked is None or self._img_w <= 0 or self._img_h <= 0:
            return 0.0, 0.0, False
        cx = (self._tracked.x + self._tracked.x2) / 2.0
        cy = (self._tracked.y + self._tracked.y2) / 2.0
        x = max(0.0, min(1.0, cx / float(self._img_w)))
        y = max(0.0, min(1.0, cy / float(self._img_h)))
        return x, y, True

    def is_present(self) -> bool:
        return self._face_present

    def miss_frames(self) -> int:
        return self._miss_frames

    def motion_bounding_rect(self) -> tuple[int, int, int, int] | None:
        return self._last_motion_rect

    def _should_scan_full_frame(self) -> bool:
        return self._tracked is None and (self._frame_counter % self._full_frame_scan_interval == 0)

    def _detect_in_region(
        self, gray: np.ndarray, rect: tuple[int, int, int, int]
    ) -> list[tuple[int, int, int, int]]:
        x, y, w, h = self._clamp_rect(rect, self._img_w, self._img_h)
        if w <= 0 or h <= 0:
            return []
        roi = gray[y : y + h, x : x + w]
        detected = self._classifier.detectMultiScale(
            roi,
            scaleFactor=1.1,
            minNeighbors=5,
            minSize=(60, 60),
        )
        return [(x + int(dx), y + int(dy), int(dw), int(dh)) for dx, dy, dw, dh in detected]

    def _push_area(self, area: int) -> None:
        if len(self._area_history) >= area_history_len:
            self._area_history.pop(0)
        self._area_history.append(area)

    def _rolling_avg_area(self) -> tuple[float, int]:
        count = len(self._area_history)
        if count == 0:
            return 0.0, 0
        return sum(self._area_history) / float(count), count

    def _is_plausible_detection(self, candidate: tuple[int, int, int, int]) -> bool:
        if len(self._area_history) < min_history_for_filter:
            return True

        avg_area, _ = self._rolling_avg_area()
        cand_area = float(candidate[2] * candidate[3])
        if cand_area > avg_area * max_area_ratio:
            return False
        if cand_area < avg_area / max_area_ratio:
            return False

        if self._tracked is not None and self._img_w > 0 and self._img_h > 0:
            diag = math.sqrt(float(self._img_w * self._img_w + self._img_h * self._img_h))
            max_jump = max_center_jump_frac * diag
            scx = (self._tracked.x + self._tracked.x2) / 2.0
            scy = (self._tracked.y + self._tracked.y2) / 2.0
            ccx = candidate[0] + candidate[2] / 2.0
            ccy = candidate[1] + candidate[3] / 2.0
            dx = ccx - scx
            dy = ccy - scy
            jump = math.sqrt(dx * dx + dy * dy)
            if jump > max_jump:
                return False

        return True

    def _motion_bounding_rect(self, mask: np.ndarray) -> tuple[int, int, int, int] | None:
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        union: tuple[int, int, int, int] | None = None
        for contour in contours:
            x, y, w, h = cv2.boundingRect(contour)
            if union is None:
                union = (int(x), int(y), int(w), int(h))
                continue
            union = self._union_rect(union, (int(x), int(y), int(w), int(h)))
        return union

    def _largest_rect(self, rects: list[tuple[int, int, int, int]]) -> tuple[int, int, int, int] | None:
        if not rects:
            return None
        return max(rects, key=lambda rect: rect[2] * rect[3])

    def _clamp_rect(self, rect: tuple[int, int, int, int], img_w: int, img_h: int) -> tuple[int, int, int, int]:
        x, y, w, h = rect
        x1 = max(0, min(img_w, x))
        y1 = max(0, min(img_h, y))
        x2 = max(x1, min(img_w, x + w))
        y2 = max(y1, min(img_h, y + h))
        return x1, y1, x2 - x1, y2 - y1

    def _union_rect(
        self, left: tuple[int, int, int, int], right: tuple[int, int, int, int]
    ) -> tuple[int, int, int, int]:
        x1 = min(left[0], right[0])
        y1 = min(left[1], right[1])
        x2 = max(left[0] + left[2], right[0] + right[2])
        y2 = max(left[1] + left[3], right[1] + right[3])
        return x1, y1, x2 - x1, y2 - y1

    def _iou_rect(
        self, left: tuple[int, int, int, int], right: tuple[int, int, int, int]
    ) -> float:
        lx1, ly1, lw, lh = left
        rx1, ry1, rw, rh = right
        lx2 = lx1 + lw
        ly2 = ly1 + lh
        rx2 = rx1 + rw
        ry2 = ry1 + rh

        inter_x1 = max(lx1, rx1)
        inter_y1 = max(ly1, ry1)
        inter_x2 = min(lx2, rx2)
        inter_y2 = min(ly2, ry2)
        if inter_x2 <= inter_x1 or inter_y2 <= inter_y1:
            return 0.0

        inter_area = float((inter_x2 - inter_x1) * (inter_y2 - inter_y1))
        left_area = float(lw * lh)
        right_area = float(rw * rh)
        union_area = left_area + right_area - inter_area
        if union_area <= 0:
            return 0.0
        return inter_area / union_area
